### ISSUE TITLE
UI: Network Manager: Swithcing between VPN connections does not update the UI

### ISSUE THREAD

**Issue by Tridy** [linuxmint/cinnamon#12550]
### Distribution

LMDE 6 Faye

### Package version

6.2.9

### Graphics hardware in use

NVIDIA Corporation GA104 [GeForce RTX 3070]

### Frequency

Always

### Bug description

In Newtork Manger when switching between VPN A and VPN B, the list of the VPNs does not get automatically updated with the currenly active VPN, and it takes one more click to update it.

### Steps to reproduce

Have 2 VPN Connections imported,
1. click on VPN A to connect (the successful connection notification comes up)
2. click on VPN B to connect (the successful connection notification comes up)
3. now VPN B is connected but the dot showing the current connection is still pointing to VPN A

additional:

4. clicking on the "VPN Connections" with the slider refreshes the list and shows the proper connection being active instead of disconnecting all VPNs.

### Expected behavior

the dot showing the currently selected VPN should move to the newly connected VPN

![vpns](https://github.com/user-attachments/assets/9cb4f68d-d3c4-48e0-ac39-e1f6e95cb87e)

**Comment by MatzFan** [comment]
Seems another dupe of #7093. On LMDE simply removing the applet was a good workaround for me. The fallback Gnome NM applet GUI can toggle any number of VPN's on and off and switch between them without issue.

**Comment by gallogabor** [comment]
Just to clarify the real problem for OP:

When you click on the second VPN connection, and it connects, the notifies you about the successful connection, that not means it disconnected automatically the first VPN you connected. So after clicking the second VPN, both VPN connections are active at the same time.
The real problem is that the network@cinnamon panel applet cannot show properly multiple active VPN connections to the user. The network@cinnamon panel applet always shows the first connected VPN, if multiple are active. And when you click the on-off switch, you disconnect the oldest connection, then the second oldest, etc.
When you click on the VPN on-off switch (which shows the first VPN's name), that disconnects the first (oldest) connected VPN, and that is why the applet then displaying the second connected VPN's name (because that will be the only active connection).
This network@cinnamon panel applet problem exists from about 8-10 years ago (I found this exact problem in 2016-2017 forum- and Reddit posts) and no solution appeared since.
Most of us affected (who have more than one VPN connection defined and switching, concurrently using them) switched back to Gnome's "factory" nm-applet, which (with a less fancy UI) properly shows all VPN connections (also in ABC order...) and users can easily toggel any of them on and off at the same time with a clear demonstration, which is active and which is not.

**Comment by Tridy** [comment]
Good point. I see that it should be possible to have 2 VPN connections connected at the same time.

In my case I have a connection that send all the traffic through VPN and a copy of the first that does only a partial routing for the resources that I need, while the rest goes as normal. I guess it would not be possible to have the connected at the same time (when a failed connection would not toggle the ON state in the UI). At this point if I am actually trying to use both at the same time, and while one is ON, and the other one is trying to connect, it makes the UI from the system tray just close when I am trying to hover the second one.

**Comment by IvanTheGeek** [comment]
There is a concrete cause for the dot not moving, separate from the multiple-active-VPN question @gallogabor raised (which is #7093).

In `NMDeviceVPN._createSection()`:

```js
if (!obj.item) {
    obj.item = new PopupMenu.PopupMenuItem(obj.name);
    this._updateConnectionItemView(obj.item, obj.connection);   // <- third argument missing
```

`_updateConnectionItemView(item, connection, active)` decides the dot from `active`. On the **first** build of a row that argument is omitted, so `active` is `undefined` and the dot is not drawn even when the connection is already up. The `else` branch further down does pass it, so the row corrects itself on the next redraw — which is exactly why this reads as "it takes one more click to update it".

#13911 includes the one-line fix. It will not by itself resolve everything in this thread, but it should remove the "needs an extra click to refresh" part.


---

**Disclosure:** I worked on this with Claude (Anthropic's Claude Code). It did the bulk of the code reading, diagnosis and patch authoring, and drafted this write-up; I reviewed the changes and ran them on my own machine before filing. The commits carry a `Co-Authored-By: Claude` trailer to the same effect. Flagging it so you can weight the review accordingly.
